README.md
# NCC Group Scrying
![Build](https://github.com/nccgroup/scrying/workflows/Build/badge.svg)
![Release](https://github.com/nccgroup/scrying/workflows/Release/badge.svg)

A new tool for collecting RDP, web and VNC screenshots all in one place

This tool is still a work-in-progress and should be mostly usable but is not yet complete.
Please file any bugs or feature requests as [GitHub issues](https://github.com/nccgroup/scrying/issues)

## Motivation
Since Eyewitness recently [dropped support for RDP](https://github.com/FortyNorthSecurity/EyeWitness/issues/422#issuecomment-539690698) there isn't a working CLI tool for capturing RDP screenshots.
Nessus still works, but it's a pain to get the images out and they're not included in the export file.

I thought this was a good opportunity to write a fresh tool that's more powerful than those that came before. Check out the feature list!

## Installation
For web screenshotting, scrying currently depends on there being an installation of Chromium or Google Chrome. Install with `pacman -S chromium` or the equivalent for your OS.

Download the latest release from [the releases tab](https://github.com/nccgroup/scrying/releases). There's a Debian package available for distros that use them (install with `sudo dpkg -i scrying*.deb`), and zipped binaries for Windows, Mac, and other Linuxes.

## Usage
Grab a single web page, RDP server, or VNC server:
```
$ scrying -t http://example.com
$ scrying -t rdp://192.0.2.1
$ scrying -t 2001:db8::5 --mode web
$ scrying -t 2001:db8::5 --mode rdp
$ scrying -t 192.0.2.2
$ scrying -t vnc://[2001:db8::53]:5901
```

Run on a headless server:
```
# apt install xvfb # or OS equivalent
$ xvfb-run scrying -t http://example.com
```

Automatically grab screenshots from an nmap output:
```
$ nmap -iL targets.txt -p 80,443,8080,8443,3389 -oX targets.xml
$ scrying --nmap targets.xml
```

Choose a different output directory for images:
```
$ scrying -t 2001:db8::3 --output-dir /tmp/scrying_outputs
```

Run from a targets file:
```
$ cat targets.txt
http://example.com
rdp://192.0.2.1
2001:db8::5
$ scrying -f targets.txt
```

Run through a web proxy:
```
$ scrying -t http://example.com --web-proxy http://127.0.0.1:8080
$ scrying -t http://example.com --web-proxy socks5://\[::1\]:1080
```

Image files are saved as PNG in the following directory structure:
```
output
├── report.html
├── rdp
│   └── 192.0.2.1-3389.png
├── vnc
│   └── 192.0.2.1-5900.png
└── web
    └── https_example.com.png
```

Check out the report at `output/report.html`!

## Features:
Features with ticks next to them have been implemented, others are TODO
* ✔️ Automatically decide whether an input should be treated as a web address or RDP server
* ✔️ Automatically create output directory if it does not already exist
* ✔️ Save images with consistent and unique filenames derived from the host/IP
* ✔️ Full support for IPv6 and IPv4 literals as well as hostnames
* ✔️ Read targets from a file and decide whether they're RDP or HTTP or use hints
* ✔️ Parse targets smartly from Nmap and Nessus output
* ✔️ HTTP - uses platform web renderer, optionally provide paths to try on each server
* ✔️ Produces an HTML report to allow easy browsing of the results
* ✔️ VNC - supports sending auth
* ✔️ RDP - mostly working, does not support "plain RDP" mode, see [#15](https://github.com/nccgroup/scrying/issues/15)
* ✔️ Customise size of captured images (web & RDP; VNC does not generally allow this)
* Proxy support - SOCKS works for RDP. Web is currently broken pending [inclusion of the set_proxy command in webkit2gtk](https://github.com/gtk-rs/webkit2gtk-rs/issues/81) [#11](https://github.com/nccgroup/scrying/issues/11)
* Video streams - tracking issue [#5](https://github.com/nccgroup/scrying/issues/5)
* option for timestamps in filenames
* Read targets from a msf services -o csv output
* OCR on RDP usernames, either live or on a directory of images
* NLA/auth to test credentials
* Parse Dirble JSON output to grab screenshots of an entire website - waiting for [nccgroup/dirble#51](https://github.com/nccgroup/dirble/issues/51)
* Full cross-platform support - tested on Linux and Windows; Mac support has been deprioritised pending good ideas for screenshotting the cocoa webview


## Help text
```
USAGE:
    scrying [OPTIONS] <--file <FILE>|--nmap <NMAP XML FILE>|--nessus <NESSUS XML FILE>|--target <TARGET>>

OPTIONS:
        --disable-report               Don't create a report.html [aliases: no-report]
    -f, --file <FILE>                  Targets file, one per line
    -h, --help                         Print help information
    -l, --log-file <LOG FILE>          Save logs to the given file
    -m, --mode <MODE>                  Force targets to be parsed as `web`, `rdp`, `vnc` [default:
                                       auto] [possible values: web, rdp, vnc, auto]
        --nessus <NESSUS XML FILE>     Nessus XML file
        --nmap <NMAP XML FILE>         Nmap XML file
    -o, --output <OUTPUT DIR>          Directory to save the captured images in [default: output]
        --proxy <PROXY>                Default SOCKS5 proxy to use for connections e.g.
                                       socks5://[::1]:1080
        --rdp-domain <RDP DOMAIN>      Domain name to provide to RDP servers that request one
        --rdp-pass <RDP PASS>          Password to provide to RDP servers that request one
        --rdp-proxy <RDP PROXY>        SOCKS5 proxy to use for RDP connections e.g.
                                       socks5://[::1]:1080
        --rdp-timeout <RDP TIMEOUT>    Seconds to wait after last bitmap before saving an image
                                       [default: 2]
        --rdp-user <RDP USER>          Username to provide to RDP servers that request one
    -s, --silent                       Suppress most log messages
        --size <SIZE>                  Set the size of captured images in pixels. Due to protocol
                                       limitations, sizes greater than 65535x65535 may get truncated
                                       in interesting ways. This argument has no effect on VNC
                                       screenshots. [default: 1280x1024]
    -t, --target <TARGET>              Target, e.g. http://example.com, rdp://[2001:db8::4]
        --test-import                  Exit after importing targets
        --threads <THREADS>            Number of worker threads for each target type [default: 10]
    -v, --verbose                      Increase log verbosity
    -V, --version                      Print version information
        --vnc-auth <VNC AUTH>          Password to provide to VNC servers that request one
        --web-mode <WEB MODE>          Choose between headless Chrom{e,ium} or native webview (GTK
                                       on Linux, Edge WebView2 on Windows, Cocoa WebView on Mac
                                       [default: chrome] [possible values: chrome, native]
        --web-path <WEB PATH>          Append a path to web requests. Provide multiple to request
                                       each path sequentially
        --web-proxy <WEB PROXY>        HTTP/SOCKS Proxy to use for web requests e.g.
                                       http://[::1]:8080
```

## Sample HTML report
![Sample report](images/scrying-report.png)


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security


## [v0.9.2] - 2022-11-07
### Fixed
* Fix Linux and macos release on CI (again)


## [v0.9.1] - 2022-11-06
### Fixed
* Fix Linux release on CI


## [v0.9.0] - 2022-11-06
### Added
* Customise the size of captured images with the `--size` option (web & RDP). Does not work on VNC because the server generally specifies the screen size [#36](https://github.com/nccgroup/scrying/issues/36)
* Optionally provide RDP credentials
* Option to skip producing a report.html [#45](https://github.com/nccgroup/scrying/issues/45)

### Changed
* Disable RDP certificate verification [#48](https://github.com/nccgroup/scrying/issues/48)

### Removed
* Removed support for platform-native webview to reduce maintenance requirements

### Fixed
* Replace question marks in URLs when generating filenames


## [v0.9.0-alpha.2] - 2021-06-22
### Added
* Add support for VNC passwords [#38](https://github.com/nccgroup/scrying/issues/38)
* Add support for requesting particular paths from web targets [#41](https://github.com/nccgroup/scrying/issues/41)

### Changed
* Log messages from the RDP and VNC modules are now tagged with the target IP [#42](https://github.com/nccgroup/scrying/issues/42)


### Removed
* Removed explicit support for macos because I can't easily test or develop for it and I don't know how best to screenshot a cocoa webview (contributions welcome if you have any ideas)


## [v0.9.0-alpha.1] - 2021-03-07

### Changed
* Windows builds now use the native Edge webview for web rendering
* Linux builds now use Webkit2GTK for web rendering
* Pressing ctrl+c once will ask the current processes to stop and still produce an output file. Pressing it again will cause Scrying to immediately exit with an error code

### Notes
* Missing proxy functionality
* Switch to native renderer on Macos is still TODO


## [v0.8.2] - 2020-11-19
### Fixed
* Debian package now depends on either `chromium`, `chromium-browser` or `google-chrome` because every Debian-derived distribution seems to have its own name for Chromium [#27](https://github.com/nccgroup/scrying/issues/27)


## [v0.8.1] - 2020-11-04
### Changed
* Enable integer overflow checks in release mode to investigate [#26](https://github.com/nccgroup/scrying/issues/26)


## [v0.8.0] - 2020-11-02

### Fixed
* Correctly parse Nmap files without full service information for each port ([nmap_xml_parser#7](https://github.com/Ayrx/nmap_xml_parser/issues/7))


## [v0.7.0] - 2020-06-29
### Added
* Added support for reading Nessus XML files
* RDP errors are collected and included in the report
* While XP is not currently supported, suspected XP-era machines have appropriate error messages [#21](https://github.com/nccgroup/scrying/issues/21)
* Catches CTRL+C so that active targets can be completed and a report produced before exiting

### Changed
* The program exits early if it couldn't parse any targets from the input files [#19](https://github.com/nccgroup/scrying/issues/19)

### Fixed
* Fix bug where connection issues with RDP would result in a panic [#22](https://github.com/nccgroup/scrying/issues/22)
* Fix bug where different input arguments would conflict [#23](https://github.com/nccgroup/scrying/issues/23)


## [v0.6.0] - 2020-06-29
### Added
* Added support for 15- and 24-bit colour depth and 8-bit colour maps on VNC

## [v0.5.0] - 2020-06-22
### Added
* Added support for VNC screenshotting [#6](https://github.com/nccgroup/scrying/issues/6)


## [v0.4.0] - 2020-06-19
### Added
* HTML report output [#7](https://github.com/nccgroup/scrying/issues/7)
* SOCKSv5 proxy support for RDP conenctions [#11](https://github.com/nccgroup/scrying/issues/11)

### Changed
* Targets are deduplicated across all input types before processing [#18](https://github.com/nccgroup/scrying/issues/18)

### Fixed
* Fixed inverted colours on RDP images
* Fixed bug where the output directory argument was ignored [#17](https://github.com/nccgroup/scrying/issues/17)

## [v0.3.0] - 2020-06-18
### Changed
* Changed project name to Scrying

## [v0.2.0] - 2020-06-17
### Added
* Implemented proxy support for web requests

### Fixed
* Fixed bug where RDP images were not received properly. [#2](https://github.com/nccgroup/scrying/issues/2)

## [v0.1.0] - 2020-06-16
### Added
* Implemented RDP screenshotting
* Implemented web screenshotting via headless Chrome
* Parse targets from Nmap XML files


src/argparse.rs
/*
 *   This file is part of NCC Group Scrying https://github.com/nccgroup/scrying
 *   Copyright 2020-2021 David Young <david(dot)young(at)nccgroup(dot)com>
 *   Released as open source by NCC Group Plc - https://www.nccgroup.com
 *
 *   Scrying is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Scrying is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with Scrying.  If not, see <https://www.gnu.org/licenses/>.
*/

use clap::{command, Arg, ArgGroup};
use color_eyre::Result;
use lazy_static::lazy_static;
use regex::Regex;
use std::str::FromStr;

lazy_static! {
    static ref SIZE_REGEX: Regex = Regex::new(r"^(\d+)x(\d+)$").unwrap();
}

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub enum Mode {
    Auto,
    Web,
    Rdp,
    Vnc,
}

impl Mode {
    /// Determine whether the supplied mode filter is valid for the
    /// current mode. Combinations are:
    /// Mode::Auto -> all filters valid
    /// Mode::X -> only X and auto are valid
    pub fn selected(&self, filter: Self) -> bool {
        use Mode::*;
        self == &Auto || self == &filter || filter == Auto
    }
}

impl Default for Mode {
    fn default() -> Self {
        Mode::Auto
    }
}

impl FromStr for Mode {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        use Mode::{Auto, Rdp, Vnc, Web};
        match s {
            "web" => Ok(Web),
            "rdp" => Ok(Rdp),
            "vnc" => Ok(Vnc),
            "auto" => Ok(Auto),
            _ => Err("Mode must be \"auto\", \"web\" or \"rdp\""),
        }
    }
}

#[derive(Debug, Default)]
pub struct Opts {
    pub files: Vec<String>,
    pub targets: Vec<String>,
    pub mode: Mode,
    pub rdp_timeout: usize,
    pub threads: usize,
    pub log_file: Option<String>,
    pub nmaps: Vec<String>,
    pub nessus: Vec<String>,
    pub output_dir: String,
    pub web_proxy: Option<String>,
    pub rdp_proxy: Option<String>,
    pub rdp_domain: Option<String>,
    pub rdp_user: Option<String>,
    pub rdp_pass: Option<String>,
    pub vnc_auth: Option<String>,
    pub web_path: Vec<String>,
    pub size: (usize, usize),
    pub silent: bool,
    pub verbose: u64,
    pub test_import: bool,
    pub disable_report: bool,
}

pub fn parse() -> Result<Opts> {
    let args = command!()
        .about("Automatic RDP, Web, and VNC screenshotting tool")
        .arg_required_else_help(true)
        .arg(
            Arg::new("FILE")
                .help("Targets file, one per line")
                .long("file")
                .multiple_occurrences(true)
                .short('f')
                .takes_value(true),
        )
        .arg(
            Arg::new("TARGET")
                .help("Target, e.g. http://example.com, rdp://[2001:db8::4]")
                .long("target")
                .multiple_occurrences(true)
                .short('t')
                .takes_value(true),
        )
        .arg(
            Arg::new("MODE")
                .help("Force targets to be parsed as `web`, `rdp`, `vnc`")
                .default_value("auto")
                .long("mode")
                .possible_values(["web", "rdp", "vnc", "auto"])
                .short('m')
                .takes_value(true),
        )
        .arg(
            Arg::new("RDP TIMEOUT")
                .help(
                    "Seconds to wait after last bitmap before saving an image",
                )
                .default_value("2")
                .long("rdp-timeout")
                .takes_value(true)
                .takes_value(true),
        )
        .arg(
            Arg::new("THREADS")
                .help("Number of worker threads for each target type")
                .default_value("10")
                .long("threads")
                .takes_value(true),
        )
        .arg(
            Arg::new("LOG FILE")
                .help("Save logs to the given file")
                .long("log-file")
                .short('l')
                .takes_value(true),
        )
        .arg(
            Arg::new("NMAP XML FILE")
                .help("Nmap XML file")
                .long("nmap")
                .multiple_occurrences(true)
                .takes_value(true),
        )
        .arg(
            Arg::new("NESSUS XML FILE")
                .help("Nessus XML file")
                .long("nessus")
                .multiple_occurrences(true)
                .takes_value(true),
        )
        .arg(
            Arg::new("OUTPUT DIR")
                .help("Directory to save the captured images in")
                .default_value("output")
                .long("output")
                .short('o')
                .takes_value(true),
        )
        .arg(
            Arg::new("DISABLE REPORT")
                .help("Don't create a report.html")
                .long("disable-report")
                .visible_alias("no-report"),
        )
        .arg(
            Arg::new("WEB PROXY")
                .help(concat!(
                    "HTTP/SOCKS Proxy to use for web requests",
                    " e.g. http://[::1]:8080"
                ))
                .long("web-proxy")
                .takes_value(true),
        )
        .arg(
            Arg::new("RDP PROXY")
                .help(concat!(
                    "SOCKS5 proxy to use for RDP connections",
                    " e.g. socks5://[::1]:1080"
                ))
                .long("rdp-proxy")
                .takes_value(true),
        )
        .arg(
            Arg::new("PROXY")
                .help(concat!(
                    "Default SOCKS5 proxy to use for connections",
                    " e.g. socks5://[::1]:1080"
                ))
                .long("proxy")
                .takes_value(true)
                .validator(is_socks5),
        )
        .arg(
            Arg::new("RDP DOMAIN")
                .help("Domain name to provide to RDP servers that request one")
                .long("rdp-domain")
                .takes_value(true),
        )
        .arg(
            Arg::new("RDP USER")
                .help("Username to provide to RDP servers that request one")
                .long("rdp-user")
                .takes_value(true),
        )
        .arg(
            Arg::new("RDP PASS")
                .help("Password to provide to RDP servers that request one")
                .long("rdp-pass")
                .takes_value(true),
        )
        .arg(
            Arg::new("VNC AUTH")
                .help("Password to provide to VNC servers that request one")
                .long("vnc-auth")
                .takes_value(true),
        )
        .arg(
            Arg::new("WEB PATH")
                .help(concat!(
                    "Append a path to web requests. Provide multiple",
                    " to request each path sequentially"
                ))
                .long("web-path")
                .multiple_occurrences(true)
                .takes_value(true),
        )
        .arg(
            Arg::new("SIZE")
                .help(concat!(
                    "Set the size of captured images in pixels.",
                    " Due to protocol limitations, sizes greater than",
                    " 65535x65535 may get truncated in interesting ways.",
                    " This argument has no effect on VNC screenshots."
                ))
                .default_value("1280x1024")
                .long("size")
                .takes_value(true)
                .validator(size_validator),
        )
        .arg(
            Arg::new("SILENT")
                .help("Suppress most log messages")
                .long("silent")
                .short('s'),
        )
        .arg(
            Arg::new("VERBOSE")
                .help("Increase log verbosity")
                .long("verbose")
                .multiple_occurrences(true)
                .short('v')
                .takes_value(false),
        )
        .arg(
            Arg::new("TEST IMPORT")
                .help("Exit after importing targets")
                .long("test-import"),
        )
        .group(
            ArgGroup::new("inputs")
                .multiple(true)
                .required(true)
                .args(&["FILE", "NMAP XML FILE", "NESSUS XML FILE", "TARGET"]),
        )
        .get_matches();

    // Grab input files if present, otherwise an empty Vec
    let mut files: Vec<String> = Vec::new();
    if let Some(f) = args.values_of("FILE") {
        for file in f {
            files.push(file.to_string());
        }
    }

    // Grab targets if present, otherwise an empty Vec
    let mut targets: Vec<String> = Vec::new();
    if let Some(t) = args.values_of("TARGET") {
        for target in t {
            targets.push(target.to_string());
        }
    }

    // Grab Nmap files if present, otherwise an empty Vec
    let mut nmaps: Vec<String> = Vec::new();
    if let Some(n) = args.values_of("NMAP XML FILE") {
        for nmap in n {
            nmaps.push(nmap.to_string());
        }
    }

    // Grab Nessus files if present, otherwise an empty Vec
    let mut nessus: Vec<String> = Vec::new();
    if let Some(n) = args.values_of("NESSUS XML FILE") {
        for nessus_file in n {
            nessus.push(nessus_file.to_string());
        }
    }

    // If global proxy setting is configured then set all indivitual
    // proxy values to it. Then override each one in turn if applicable
    let mut web_proxy = None;
    let mut rdp_proxy = None;
    if let Some(p) = args.value_of("PROXY") {
        web_proxy = Some(p.to_string());
        rdp_proxy = Some(p.to_string());
    }

    if let Some(p) = args.value_of("RDP PROXY") {
        rdp_proxy = Some(p.to_string());
    }

    if let Some(p) = args.value_of("WEB PROXY") {
        web_proxy = Some(p.to_string());
    }

    // SIZE has already passed the regex validator, so the following
    // extraction probably won't panic
    let captures = SIZE_REGEX
        .captures_iter(args.value_of("SIZE").unwrap())
        .next()
        .unwrap();
    let size = (captures[1].parse().unwrap(), captures[2].parse().unwrap());

    Ok(Opts {
        files,
        targets,
        mode: args.value_of_t("MODE").unwrap(),
        rdp_timeout: args.value_of_t("RDP TIMEOUT").unwrap(),
        threads: args.value_of_t("THREADS").unwrap(),
        log_file: args.value_of("LOG FILE").map(String::from),
        nmaps,
        nessus,
        output_dir: args.value_of_t("OUTPUT DIR").unwrap(),
        web_proxy,
        rdp_proxy,
        vnc_auth: args.value_of("VNC AUTH").map(String::from),
        rdp_domain: args.value_of("RDP DOMAIN").map(String::from),
        rdp_user: args.value_of("RDP USER").map(String::from),
        rdp_pass: args.value_of("RDP PASS").map(String::from),
        web_path: if let Some(paths) = args.values_of("WEB PATH") {
            paths.map(String::from).collect()
        } else {
            Vec::new()
        },
        size,
        silent: args.is_present("SILENT"),
        verbose: args.occurrences_of("VERBOSE"),
        test_import: args.is_present("TEST IMPORT"),
        disable_report: args.is_present("DISABLE REPORT"),
    })
}

fn is_socks5(val: &str) -> Result<(), String> {
    if !val.starts_with("socks5://") {
        Err("Global or RDP proxy must be a socks5:// URI".to_string())
    } else {
        Ok(())
    }
}

fn size_validator(val: &str) -> Result<(), String> {
    if SIZE_REGEX.is_match(val) {
        Ok(())
    } else {
        Err(r#"Size must be in the form "1280x720" (regex "^\d+x\d+$""#
            .to_string())
    }
}

#[cfg(test)]
mod test {
    #[test]
    fn mode_filter() {
        use super::Mode::*;

        let auto = Auto;
        let rdp = Rdp;
        let web = Web;

        assert!(auto.selected(Auto));
        assert!(auto.selected(Rdp));
        assert!(auto.selected(Web));

        assert!(rdp.selected(Auto));
        assert!(rdp.selected(Rdp));
        assert!(!rdp.selected(Web));

        assert!(web.selected(Auto));
        assert!(!web.selected(Rdp));
        assert!(web.selected(Web));
    }
}
